compiler/rustc_hir_analysis/src/delegation.rs

```diff
@@ -0,0 +1,242 @@
+            let mut own_params = vec![];
+
+            let callee_generics = tcx.generics_of(sig_id);
+            if let Some(parent_sig_id) = callee_generics.parent {
+                let parent_sig_generics = tcx.generics_of(parent_sig_id);
+                own_params.append(&mut parent_sig_generics.own_params.clone());
+            }
+            own_params.append(&mut callee_generics.own_params.clone());
+
+            // lifetimes go first
+            own_params.sort_by_key(|key| key.kind.is_ty_or_const());
+
+            for (idx, param) in own_params.iter_mut().enumerate() {
+                param.index = idx as u32;
+                // Default type parameters are not inherited: they are not allowed
+                // in fn's.
+                if let ty::GenericParamDefKind::Type { synthetic, .. } = param.kind {
+                    param.kind = ty::GenericParamDefKind::Type { has_default: false, synthetic }
+                }
+            }
+
+            let param_def_id_to_index =
+                own_params.iter().map(|param| (param.def_id, param.index)).collect();
+
+            Some(ty::Generics {
+                parent: None,
+                parent_count: 0,
+                own_params,
+                param_def_id_to_index,
+                has_self: false,
+                has_late_bound_regions: callee_generics.has_late_bound_regions,
+                host_effect_index: None,
+            })
+        }
+        _ => None,
+    }
+}
+
+pub(crate) fn inherit_predicates_for_delegation_item<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    def_id: LocalDefId,
+    sig_id: DefId,
+) -> Option<ty::GenericPredicates<'tcx>> {
+    let caller_kind = fn_kind(tcx, def_id.into());
+    let callee_kind = fn_kind(tcx, sig_id);
+
+    // FIXME(fn_delegation): Support generics on associated delegation items.
+    // Error was reported in `check_delegation_constraints`.
+    match (caller_kind, callee_kind) {
+        (FnKind::Free, _) => {
+            let mut predicates = vec![];
+            let callee_predicates = tcx.predicates_of(sig_id);
+            if let Some(parent_sig_id) = callee_predicates.parent {
+                let parent_sig_predicates = tcx.predicates_of(parent_sig_id);
+                predicates.extend_from_slice(parent_sig_predicates.predicates);
+            }
+            predicates.extend_from_slice(callee_predicates.predicates);
+
+            let remap_table = create_remap_table(tcx, def_id, sig_id);
+            let mut folder = IndicesFolder { tcx, remap_table: &remap_table };
+            let predicates = predicates.fold_with(&mut folder);
+
+            Some(ty::GenericPredicates {
+                parent: None,
+                predicates: tcx.arena.alloc_from_iter(predicates),
+            })
+        }
+        _ => None,
+    }
+}
+
+fn check_constraints<'tcx>(tcx: TyCtxt<'tcx>, def_id: LocalDefId) -> Result<(), ErrorGuaranteed> {
+    let mut ret = Ok(());
+
+    let span = tcx.def_span(def_id);
```

Comment:
Can move the `def_span` calls into the closure since they are only used on the error path.